Panic observed if a malformed wire message is sent

Node panics when frame size field has values in the interval  [5,7] due to lack of len check in `checksum.Extract` 

To reproduce:

```go
        frame := new(bytes.Buffer)
	// Add length bytes
	if err := encoding.WriteUint64LE(frame, 7); err != nil {
		return err
	}

        // Add correct magic value
	magic := protocol.TestNet
	mBuf := magic.ToBuffer()
	if _, err := frame.Write(mBuf.Bytes()); err != nil {
		return err
	}
       
       // Checksum with 3 bytes only
       if _, err := frame.Write([]byte{0, 0, 0}); err != nil {
			t.Error(err)
       }
        // malformed payload ( 3 bytes-only checksum, 0 message )
	_, err := c.Write(frame.Bytes())
```

```go
panic: runtime error: slice bounds out of range [:4] with capacity 3

goroutine 37 [running]:
github.com/dusk-network/dusk-blockchain/pkg/p2p/wire/checksum.Extract(...)
       dusk-blockchain/pkg/p2p/wire/checksum/checksum.go:18
github.com/dusk-network/dusk-blockchain/pkg/p2p/peer.(*Reader).ReadLoop(0xc0003002e0)
      dusk-blockchain/pkg/p2p/peer/peer.go:237 +0x6cf
created by main.(*Server).OnAccept
       dusk-blockchain/launch/testnet/server.go:147 +0x553
```

--- a/pkg/p2p/peer/peer.go
+++ b/pkg/p2p/peer/peer.go
@@ -242,7 +242,7 @@ func (p *Reader) readLoop() {
 		// Refresh the read deadline
 		err := p.Conn.SetReadDeadline(time.Now().Add(readWriteTimeout))
 		if err != nil {
-			l.WithError(err).Warnf("write timeout err: %v", err)
+			l.WithError(err).Warnf("error setting read timeout")
 		}
 
 		b, err := p.ReadMessage()

--- a/pkg/p2p/peer/peer.go
+++ b/pkg/p2p/peer/peer.go
@@ -213,6 +213,20 @@ func (w *Writer) writeLoop(writeQueueChan <-chan *bytes.Buffer, exitChan chan st
 // is reached. Should be called in a go-routine, after a successful handshake with
 // a peer. Eventual duplicated messages are silently discarded.
 func (p *Reader) ReadLoop() {
+
+	// As the peer ReadLoop is at the front-line of P2P network, receiving a
+	// malformed frame by an adversary node could lead to a panic.
+	// In such situation, the node should survive but adversary conn gets dropped
+	defer func() {
+		if r := recover(); r != nil {
+			log.Errorf("Peer %s failed with critical issue: %v", p.RemoteAddr(), r)
+		}
+	}()
+
+	p.readLoop()
+}
+
+func (p *Reader) readLoop() {
 	defer p.Conn.Close()
 
 	// Set up a timer, which triggers the sending of a `keepalive` message
@@ -226,7 +240,10 @@ func (p *Reader) ReadLoop() {
 
 	for {
 		// Refresh the read deadline
-		p.Conn.SetReadDeadline(time.Now().Add(readWriteTimeout))
+		err := p.Conn.SetReadDeadline(time.Now().Add(readWriteTimeout))
+		if err != nil {
+			l.WithError(err).Warnf("write timeout err: %v", err)
+		}
 
 		b, err := p.ReadMessage()
 		if err != nil {

--- a/pkg/p2p/peer/peer_in_test.go
+++ b/pkg/p2p/peer/peer_in_test.go
@@ -9,10 +9,12 @@ import (
 	"github.com/dusk-network/dusk-blockchain/pkg/p2p/peer/dupemap"
 	"github.com/dusk-network/dusk-blockchain/pkg/p2p/peer/processing"
 	"github.com/dusk-network/dusk-blockchain/pkg/p2p/peer/processing/chainsync"
+	"github.com/dusk-network/dusk-blockchain/pkg/p2p/wire/encoding"
 	"github.com/dusk-network/dusk-blockchain/pkg/p2p/wire/protocol"
 	"github.com/dusk-network/dusk-blockchain/pkg/p2p/wire/topics"
 	"github.com/dusk-network/dusk-blockchain/pkg/util/nativeutils/eventbus"
 	"github.com/dusk-network/dusk-blockchain/pkg/util/nativeutils/rpcbus"
+	"github.com/stretchr/testify/assert"
 )
 
 // Test that the 'ping' message is sent correctly, and that a 'pong' message will result.
@@ -54,4 +56,59 @@ func TestPingLoop(t *testing.T) {
 			break
 		}
 	}
+}
+
+// TestMalformedFrame ensures peer reader does not panic on
+// incomplete checksum buffer
+func TestMalformedFrame(t *testing.T) {
+
+	fn := func() {
+		peer, r, w, _ := dummyReader()
+		defer r.Close()
+		defer w.Close()
+
+		// Run the non-recover readLoop to watch for panics
+		go peer.readLoop()
+
+		time.Sleep(time.Second)
+
+		// Construct an invalid frame
+		frame := new(bytes.Buffer)
+		// Add length bytes
+		if err := encoding.WriteUint64LE(frame, 7); err != nil {
+			t.Error(err)
+		}
+
+		// Add correct magic value
+		mBuf := protocol.TestNet.ToBuffer()
+		if _, err := frame.Write(mBuf.Bytes()); err != nil {
+			t.Error(err)
+		}
+
+		// Checksum with 3 bytes only
+		if _, err := frame.Write([]byte{0, 0, 0}); err != nil {
+			t.Error(err)
+		}
+
+		_, _ = w.Write(frame.Bytes())
+		time.Sleep(1 * time.Second)
+	}
+
+	assert.NotPanics(t, fn)
+}
+
+func dummyReader() (*Reader, net.Conn, net.Conn, chan<- *bytes.Buffer) {
+
+	bus := eventbus.New()
+	rpcbus := rpcbus.New()
+	d := dupemap.NewDupeMap(0)
+	r, w := net.Pipe()
+	keepAliveTime = 30 * time.Second
+
+	respChan := make(chan *bytes.Buffer, 10)
+	g := processing.NewGossip(protocol.TestNet)
+	peer, _ := NewReader(r, g, d, bus, rpcbus, &chainsync.Counter{},
+		respChan, make(chan struct{}, 1))
+
+	return peer, r, w, respChan
 }

--- a/pkg/p2p/peer/peer_in_test.go
+++ b/pkg/p2p/peer/peer_in_test.go
@@ -96,6 +96,37 @@ func TestMalformedFrame(t *testing.T) {
 	assert.NotPanics(t, fn)
 }
 
+func TestZeroLength(t *testing.T) {
+
+	fn := func() {
+		peer, _, w, _ := dummyReader()
+		defer peer.Close()
+
+		// Run the non-recover readLoop to watch for panics
+		go peer.readLoop()
+
+		time.Sleep(time.Second)
+
+		// Construct an invalid frame
+		frame := new(bytes.Buffer)
+		// Add 0 length (any value smaller than 4)
+		if err := encoding.WriteUint64LE(frame, 0); err != nil {
+			t.Error(err)
+		}
+
+		// Add correct magic value
+		mBuf := protocol.TestNet.ToBuffer()
+		if _, err := frame.Write(mBuf.Bytes()); err != nil {
+			t.Error(err)
+		}
+
+		_, _ = w.Write(frame.Bytes())
+		time.Sleep(1 * time.Second)
+	}
+
+	assert.NotPanics(t, fn)
+}
+
 func dummyReader() (*Reader, net.Conn, net.Conn, chan<- *bytes.Buffer) {
 
 	bus := eventbus.New()

--- a/pkg/p2p/peer/peer_in_test.go
+++ b/pkg/p2p/peer/peer_in_test.go
@@ -9,6 +9,7 @@ import (
 	"github.com/dusk-network/dusk-blockchain/pkg/p2p/peer/dupemap"
 	"github.com/dusk-network/dusk-blockchain/pkg/p2p/peer/processing"
 	"github.com/dusk-network/dusk-blockchain/pkg/p2p/peer/processing/chainsync"
+	"github.com/dusk-network/dusk-blockchain/pkg/p2p/wire/checksum"
 	"github.com/dusk-network/dusk-blockchain/pkg/p2p/wire/encoding"
 	"github.com/dusk-network/dusk-blockchain/pkg/p2p/wire/protocol"
 	"github.com/dusk-network/dusk-blockchain/pkg/p2p/wire/topics"
@@ -58,76 +59,116 @@ func TestPingLoop(t *testing.T) {
 	}
 }
 
-// TestMalformedFrame ensures peer reader does not panic on
+// TestIncompleteChecksum ensures peer reader does not panic on
 // incomplete checksum buffer
-func TestMalformedFrame(t *testing.T) {
+func TestIncompleteChecksum(t *testing.T) {
 
-	fn := func() {
-		peer, _, w, _ := dummyReader()
-		defer peer.Close()
+	peer, _, w, _ := testReader(t)
+	defer peer.Close()
 
-		// Run the non-recover readLoop to watch for panics
-		go peer.readLoop()
-
-		time.Sleep(time.Second)
-
-		// Construct an invalid frame
-		frame := new(bytes.Buffer)
-		// Add length bytes
-		if err := encoding.WriteUint64LE(frame, 7); err != nil {
-			t.Error(err)
-		}
-
-		// Add correct magic value
-		mBuf := protocol.TestNet.ToBuffer()
-		if _, err := frame.Write(mBuf.Bytes()); err != nil {
-			t.Error(err)
-		}
-
-		// Checksum with 3 bytes only
-		if _, err := frame.Write([]byte{0, 0, 0}); err != nil {
-			t.Error(err)
-		}
-
-		_, _ = w.Write(frame.Bytes())
-		time.Sleep(1 * time.Second)
+	// Construct an invalid frame
+	frame := new(bytes.Buffer)
+	// Add length bytes
+	if err := encoding.WriteUint64LE(frame, 7); err != nil {
+		t.Error(err)
 	}
 
-	assert.NotPanics(t, fn)
+	// Add correct magic value
+	mBuf := protocol.TestNet.ToBuffer()
+	if _, err := frame.Write(mBuf.Bytes()); err != nil {
+		t.Error(err)
+	}
+
+	// Checksum with 3 bytes only
+	if _, err := frame.Write([]byte{0, 0, 0}); err != nil {
+		t.Error(err)
+	}
+
+	_, err := w.Write(frame.Bytes())
+	if err != nil {
+		t.Error(err)
+	}
+
+	time.Sleep(100 * time.Millisecond)
 }
 
+// TestZeroLength ensures peer reader does not panic on 0 length field
 func TestZeroLength(t *testing.T) {
 
-	fn := func() {
-		peer, _, w, _ := dummyReader()
-		defer peer.Close()
+	peer, _, w, _ := testReader(t)
+	defer peer.Close()
 
-		// Run the non-recover readLoop to watch for panics
-		go peer.readLoop()
-
-		time.Sleep(time.Second)
-
-		// Construct an invalid frame
-		frame := new(bytes.Buffer)
-		// Add 0 length (any value smaller than 4)
-		if err := encoding.WriteUint64LE(frame, 0); err != nil {
-			t.Error(err)
-		}
-
-		// Add correct magic value
-		mBuf := protocol.TestNet.ToBuffer()
-		if _, err := frame.Write(mBuf.Bytes()); err != nil {
-			t.Error(err)
-		}
-
-		_, _ = w.Write(frame.Bytes())
-		time.Sleep(1 * time.Second)
+	// Construct an invalid frame
+	frame := new(bytes.Buffer)
+	// Add 0 length (any value smaller than 4)
+	if err := encoding.WriteUint64LE(frame, 0); err != nil {
+		t.Error(err)
 	}
 
-	assert.NotPanics(t, fn)
+	// Add correct magic value
+	mBuf := protocol.TestNet.ToBuffer()
+	if _, err := frame.Write(mBuf.Bytes()); err != nil {
+		t.Error(err)
+	}
+
+	_, err := w.Write(frame.Bytes())
+	if err != nil {
+		t.Error(err)
+	}
+
+	time.Sleep(100 * time.Millisecond)
 }
 
-func dummyReader() (*Reader, net.Conn, net.Conn, chan<- *bytes.Buffer) {
+// TestOverflowLength
+// Ensure peer reader does not panic overflow length value
+func TestOverflowLength(t *testing.T) {
+
+	peer, _, w, _ := testReader(t)
+	defer peer.Close()
+
+	// Construct an invalid frame with large length field
+	frame := new(bytes.Buffer)
+	if err := encoding.WriteUint64LE(frame, 999999999); err != nil {
+		t.Error(err)
+	}
+
+	_, err := w.Write(frame.Bytes())
+	if err != nil {
+		t.Error(err)
+	}
+
+	time.Sleep(100 * time.Millisecond)
+}
+
+// TestInvalidPayload
+// Ensure peer reader does not panic on sending invalid payload to any topic
+func TestInvalidPayload(t *testing.T) {
+
+	peer, _, w, _ := testReader(t)
+	defer peer.Close()
+
+	// Send for all possible topic values with a malformed
+	// payload
+	var topic byte
+	for topic = 0; topic <= 50; topic++ {
+
+		buf := &bytes.Buffer{}
+		buf.WriteByte(topic)
+		buf.Write([]byte{0, 1, 2})
+
+		cs := checksum.Generate(buf.Bytes())
+		_ = processing.WriteFrame(buf, protocol.TestNet, cs)
+
+		_, err := w.Write(buf.Bytes())
+		if err != nil {
+			t.Error(err)
+		}
+
+		time.Sleep(20 * time.Millisecond)
+	}
+}
+
+func testReader(t *testing.T) (*Reader, net.Conn, net.Conn, chan<- *bytes.Buffer) {
 
 	bus := eventbus.New()
 	rpcbus := rpcbus.New()
@@ -138,6 +179,11 @@ func dummyReader() (*Reader, net.Conn, net.Conn, chan<- *bytes.Buffer) {
 	g := processing.NewGossip(protocol.TestNet)
 	peer, _ := NewReader(r, g, d, bus, rpcbus, &chainsync.Counter{},
 		respChan, make(chan struct{}, 1))
+
+	// Run the non-recover readLoop to watch for panics
+	go assert.NotPanics(t, peer.readLoop)
+
+	time.Sleep(200 * time.Millisecond)
 
 	return peer, r, w, respChan
 }

--- a/pkg/p2p/peer/peer_in_test.go
+++ b/pkg/p2p/peer/peer_in_test.go
@@ -63,9 +63,8 @@ func TestPingLoop(t *testing.T) {
 func TestMalformedFrame(t *testing.T) {
 
 	fn := func() {
-		peer, r, w, _ := dummyReader()
-		defer r.Close()
-		defer w.Close()
+		peer, _, w, _ := dummyReader()
+		defer peer.Close()
 
 		// Run the non-recover readLoop to watch for panics
 		go peer.readLoop()
@@ -103,7 +102,6 @@ func dummyReader() (*Reader, net.Conn, net.Conn, chan<- *bytes.Buffer) {
 	rpcbus := rpcbus.New()
 	d := dupemap.NewDupeMap(0)
 	r, w := net.Pipe()
-	keepAliveTime = 30 * time.Second
 
 	respChan := make(chan *bytes.Buffer, 10)
 	g := processing.NewGossip(protocol.TestNet)

--- a/pkg/p2p/peer/processing/gossip.go
+++ b/pkg/p2p/peer/processing/gossip.go
@@ -3,6 +3,7 @@ package processing
 import (
 	"bytes"
 	"errors"
+	"fmt"
 	"io"
 
 	"github.com/dusk-network/dusk-blockchain/pkg/p2p/wire/checksum"
@@ -46,5 +47,13 @@ func (g *Gossip) UnpackLength(r io.Reader) (uint64, error) {
 		return 0, errors.New("magic mismatch")
 	}
 
-	return packetLength - uint64(magic.Len()), nil
+	// If packetLength is less than magic.Len(), ln is close to MaxUint64
+	// due to integer overflow
+	ln := packetLength - uint64(magic.Len())
+
+	if ln > MaxFrameSize {
+		return 0, fmt.Errorf("invalid packet length %d", packetLength)
+	}
+
+	return ln, nil
 }

--- a/pkg/p2p/peer/responding/blockhashbroker.go
+++ b/pkg/p2p/peer/responding/blockhashbroker.go
@@ -2,6 +2,7 @@ package responding
 
 import (
 	"bytes"
+	"errors"
 
 	"github.com/dusk-network/dusk-blockchain/pkg/core/database"
 	"github.com/dusk-network/dusk-blockchain/pkg/p2p/peer/peermsg"
@@ -79,6 +80,11 @@ func (b *BlockHashBroker) AdvertiseMissingBlocks(m *bytes.Buffer) error {
 
 // Determine a peer's height from his locator hash.
 func (b *BlockHashBroker) fetchLocatorHeight(msg *peermsg.GetBlocks) (uint64, error) {
+
+	if len(msg.Locators) == 0 {
+		return 0, errors.New("empty locators array")
+	}
+
 	var height uint64
 	err := b.db.View(func(t database.Transaction) error {
 		header, err := t.FetchBlockHeader(msg.Locators[0])

--- a/pkg/p2p/wire/checksum/checksum.go
+++ b/pkg/p2p/wire/checksum/checksum.go
@@ -2,6 +2,7 @@ package checksum
 
 import (
 	"bytes"
+	"errors"
 
 	"golang.org/x/crypto/blake2b"
 )
@@ -15,6 +16,10 @@ func Generate(m []byte) []byte {
 
 func Extract(m []byte) ([]byte, []byte, error) {
 	// First 4 bytes are the checksum
+	if len(m) < Length {
+		return nil, nil, errors.New("invalid buffer size")
+	}
+
 	checksum := m[:Length]
 	message := m[Length:]
 	return message, checksum, nil

--- a/pkg/p2p/wire/checksum/checksum_test.go
+++ b/pkg/p2p/wire/checksum/checksum_test.go
@@ -3,7 +3,9 @@ package checksum_test
 import (
 	"testing"
 
+	"github.com/dusk-network/dusk-blockchain/pkg/p2p/wire/checksum"
 	crypto "github.com/dusk-network/dusk-crypto/hash"
+	"github.com/stretchr/testify/assert"
 	"golang.org/x/crypto/blake2b"
 	"golang.org/x/crypto/sha3"
 )
@@ -24,5 +26,18 @@ func BenchmarkBlake2b(b *testing.B) {
 	message, _ := crypto.RandEntropy(200)
 	for i := 0; i < b.N; i++ {
 		_ = blake2b.Sum256(message)
+	}
+}
+
+func TestVerify(t *testing.T) {
+	message, _ := crypto.RandEntropy(32)
+	cs := checksum.Generate(message)
+	assert.True(t, checksum.Verify(message, cs))
+}
+
+func TestInvalidLen(t *testing.T) {
+	_, _, err := checksum.Extract([]byte{0, 0, 0})
+	if err == nil {
+		t.Errorf("invalid len err expected")
 	}
 }

